File: solutions/java/min-heap/step06_size/MinHeap.java
import java.util.ArrayList;
import java.util.List;

public class MinHeap {
    private final List<Integer> items;

    public MinHeap() {
        this.items = new ArrayList<>();
    }

    public void insert(int value) {
        items.add(value);
        bubbleUp(items.size() - 1);
    }

    public Integer peekMin() {
        if (isEmpty()) {
            return null;
        }
        return items.get(0);
    }

    public Integer extractMin() {
        if (isEmpty()) {
            return null;
        }
        int minimum = items.get(0);
        int end = items.remove(items.size() - 1);
        if (!isEmpty()) {
            items.set(0, end);
            bubbleDown(0);
        }
        return minimum;
    }

    public int size() {
        return items.size();
    }

    public boolean isEmpty() {
        return items.isEmpty();
    }

    private void bubbleUp(int index) {
        int current = index;
        while (current > 0) {
            int parent = (current - 1) / 2;
            if (items.get(parent) <= items.get(current)) {
                break;
            }
            swap(parent, current);
            current = parent;
        }
    }

    private void bubbleDown(int index) {
        int current = index;
        int length = items.size();
        while (true) {
            int left = current * 2 + 1;
            int right = current * 2 + 2;
            int smallest = current;

            if (left < length && items.get(left) < items.get(smallest)) {
                smallest = left;
            }
            if (right < length && items.get(right) < items.get(smallest)) {
                smallest = right;
            }
            if (smallest == current) {
                break;
            }
            swap(current, smallest);
            current = smallest;
        }
    }

    private void swap(int i, int j) {
        int temp = items.get(i);
        items.set(i, items.get(j));
        items.set(j, temp);
    }
}

File: solutions/python/min-heap/05_heapify_down/min_heap.py
class MinHeap:
    def __init__(self) -> None:
        self._items = []

    def insert(self, value) -> None:
        self._items.append(value)
        self._bubble_up(len(self._items) - 1)

    def peek_min(self):
        if not self._items:
            return None
        return self._items[0]

    def extract_min(self):
        if not self._items:
            return None
        minimum = self._items[0]
        end = self._items.pop()
        if self._items:
            self._items[0] = end
            self._bubble_down(0)
        return minimum

    def _bubble_up(self, index: int) -> None:
        current = index
        while current > 0:
            parent = (current - 1) // 2
            if self._items[parent] <= self._items[current]:
                break
            # Swap parent and current
            temp = self._items[current]
            self._items[current] = self._items[parent]
            self._items[parent] = temp
            current = parent

    def _bubble_down(self, index: int) -> None:
        length = len(self._items)
        current = index
        while True:
            left = current * 2 + 1
            right = current * 2 + 2
            smallest = current

            if left < length and self._items[left] < self._items[smallest]:
                smallest = left
            if right < length and self._items[right] < self._items[smallest]:
                smallest = right
            if smallest == current:
                break
            # Swap current and smallest
            temp = self._items[current]
            self._items[current] = self._items[smallest]
            self._items[smallest] = temp
            current = smallest

